File: referralhero/src/main/java/com/sdk/rh/Logger.java
package com.sdk.rh;

import android.util.Log;

import java.util.Arrays;
import java.util.Locale;

public class Logger implements ILogger {

    private static final String formatErrorMessage = "Error formating log message: %s, with params: %s";
    String LOGTAG = "ReferralHero";
    private LogLevel logLevel;
    private boolean logLevelLocked;
    private boolean isProductionEnvironment;

    public Logger() {
        isProductionEnvironment = false;
        logLevelLocked = false;
        setLogLevel(LogLevel.INFO, isProductionEnvironment);
    }

    @Override
    public void setLogLevel(LogLevel logLevel, boolean isProductionEnvironment) {
        if (logLevelLocked) {
            return;
        }
        this.logLevel = logLevel;
        this.isProductionEnvironment = isProductionEnvironment;
    }

    @Override
    public void setLogLevelString(String logLevelString, boolean isProductionEnvironment) {
        if (null != logLevelString) {
            try {
                setLogLevel(LogLevel.valueOf(logLevelString.toUpperCase(Locale.US)), isProductionEnvironment);
            } catch (IllegalArgumentException iae) {
                error("Malformed logLevel '%s', falling back to 'info'", logLevelString);
            }
        }
    }

    @Override
    public void verbose(String message, Object... parameters) {
        if (isProductionEnvironment) {
            return;
        }
        if (logLevel.androidLogLevel <= Log.VERBOSE) {
            try {
                Log.v(LOGTAG, RHUtil.INSTANCE.formatString(message, parameters));
            } catch (Exception e) {
                Log.e(LOGTAG, RHUtil.INSTANCE.formatString(formatErrorMessage, message, Arrays.toString(parameters)));
            }
        }
    }

    @Override
    public void debug(String message, Object... parameters) {
        if (isProductionEnvironment) {
            return;
        }
        if (logLevel.androidLogLevel <= Log.DEBUG) {
            try {
                Log.d(LOGTAG, RHUtil.INSTANCE.formatString(message, parameters));
            } catch (Exception e) {
                Log.e(LOGTAG, RHUtil.INSTANCE.formatString(formatErrorMessage, message, Arrays.toString(parameters)));
            }
        }
    }

    @Override
    public void info(String message, Object... parameters) {
        if (isProductionEnvironment) {
            return;
        }
        if (logLevel.androidLogLevel <= Log.INFO) {
            try {
                Log.i(LOGTAG, RHUtil.INSTANCE.formatString(message, parameters));
            } catch (Exception e) {
                Log.e(LOGTAG, RHUtil.INSTANCE.formatString(formatErrorMessage, message, Arrays.toString(parameters)));
            }
        }
    }

    @Override
    public void warn(String message, Object... parameters) {
        if (isProductionEnvironment) {
            return;
        }
        if (logLevel.androidLogLevel <= Log.WARN) {
            try {
                Log.w(LOGTAG, RHUtil.INSTANCE.formatString(message, parameters));
            } catch (Exception e) {
                Log.e(LOGTAG, RHUtil.INSTANCE.formatString(formatErrorMessage, message, Arrays.toString(parameters)));
            }
        }
    }

    @Override
    public void warnInProduction(String message, Object... parameters) {
        if (logLevel.androidLogLevel <= Log.WARN) {
            try {
                Log.w(LOGTAG, RHUtil.INSTANCE.formatString(message, parameters));
            } catch (Exception e) {
                Log.e(LOGTAG, RHUtil.INSTANCE.formatString(formatErrorMessage, message, Arrays.toString(parameters)));
            }
        }
    }


    @Override
    public void error(String message, Object... parameters) {
        if (isProductionEnvironment) {
            return;
        }
        if (logLevel.androidLogLevel <= Log.ERROR) {
            try {
                Log.e(LOGTAG, RHUtil.INSTANCE.formatString(message, parameters));
            } catch (Exception e) {
                Log.e(LOGTAG, RHUtil.INSTANCE.formatString(formatErrorMessage, message, Arrays.toString(parameters)));
            }
        }
    }

    @Override
    public void Assert(String message, Object... parameters) {
        if (isProductionEnvironment) {
            return;
        }
        if (logLevel.androidLogLevel <= Log.ASSERT) {
            try {
                Log.println(Log.ASSERT, LOGTAG, RHUtil.INSTANCE.formatString(message, parameters));
            } catch (Exception e) {
                Log.e(LOGTAG, RHUtil.INSTANCE.formatString(formatErrorMessage, message, Arrays.toString(parameters)));
            }
        }
    }

    @Override
    public void lockLogLevel() {
        logLevelLocked = true;
    }
}

File: referralhero/src/main/java/com/sdk/referral/logger/Logger.kt
package com.sdk.referral.logger

import android.util.Log
import java.util.Locale

class Logger : ILogger {
    var LOGTAG = "ReferralHero SDK:"
    private var logLevel: LogLevel? = null
    private var logLevelLocked = false
    private var isProductionEnvironment = false

    init {
        setLogLevel(LogLevel.INFO, isProductionEnvironment)
    }

    override fun setLogLevel(logLevel: LogLevel?, isProductionEnvironment: Boolean) {
        if (logLevelLocked) {
            return
        }
        this.logLevel = logLevel
        this.isProductionEnvironment = isProductionEnvironment
    }

    override fun setLogLevelString(logLevelString: String?, isProductionEnvironment: Boolean) {
        if (null != logLevelString) {
            try {
                setLogLevel(LogLevel.valueOf(logLevelString.uppercase()), isProductionEnvironment)
            } catch (iae: IllegalArgumentException) {
                error("Malformed logLevel '%s', falling back to 'info'", logLevelString)
            }
        }
    }

    override fun verbose(message: String?, vararg parameters: Any?) {
        if (isProductionEnvironment) {
            return
        }
        if ((logLevel?.androidLogLevel ?: Log.VERBOSE) <= Log.VERBOSE) {
            try {
                formatString(message, parameters)?.let { Log.v(LOGTAG, it) }
            } catch (e: Exception) {
                formatString(formatErrorMessage, message, parameters.contentToString())?.let {
                    Log.e(
                        LOGTAG,
                        it
                    )
                }
            }
        }
    }

    override fun debug(message: String?, vararg parameters: Any?) {
        if (isProductionEnvironment) {
            return
        }
        if ((logLevel?.androidLogLevel ?: Log.DEBUG) <= Log.DEBUG) {
            try {
                formatString(message, parameters)?.let { Log.d(LOGTAG, it) }
            } catch (e: Exception) {
                formatString(formatErrorMessage, message, parameters.contentToString())?.let {
                    Log.e(
                        LOGTAG,
                        it
                    )
                }
            }
        }
    }

    override fun info(message: String?, vararg parameters: Any?) {
        if (isProductionEnvironment) {
            return
        }
        if ((logLevel?.androidLogLevel ?: Log.INFO) <= Log.INFO) {
            try {
                formatString(message, parameters)?.let { Log.i(LOGTAG, it) }
            } catch (e: Exception) {
                formatString(formatErrorMessage, message, parameters.contentToString())?.let {
                    Log.e(
                        LOGTAG,
                        it
                    )
                }
            }
        }
    }

    override fun warn(message: String?, vararg parameters: Any?) {
        if (isProductionEnvironment) {
            return
        }
        if ((logLevel?.androidLogLevel ?: Log.WARN) <= Log.WARN) {
            try {
                formatString(message, parameters)?.let { Log.w(LOGTAG, it) }
            } catch (e: Exception) {
                formatString(formatErrorMessage, message, parameters.contentToString())?.let {
                    Log.e(
                        LOGTAG,
                        it
                    )
                }
            }
        }
    }

    override fun warnInProduction(message: String?, vararg parameters: Any?) {
        if ((logLevel?.androidLogLevel ?: Log.WARN) <= Log.WARN) {
            try {
                formatString(message, parameters)?.let { Log.w(LOGTAG, it) }
            } catch (e: Exception) {
                formatString(formatErrorMessage, message, parameters.contentToString())?.let {
                    Log.e(
                        LOGTAG,
                        it
                    )
                }
            }
        }
    }

    override fun error(message: String?, vararg parameters: Any?) {
        if (isProductionEnvironment) {
            return
        }
        if ((logLevel?.androidLogLevel ?: Log.ERROR) <= Log.ERROR) {
            try {
                formatString(message, parameters)?.let { Log.e(LOGTAG, it) }
            } catch (e: Exception) {
                formatString(formatErrorMessage, message, parameters.contentToString())?.let {
                    Log.e(
                        LOGTAG,
                        it
                    )
                }
            }
        }
    }

    override fun Assert(message: String?, vararg parameters: Any?) {
        if (isProductionEnvironment) {
            return
        }
        if ((logLevel?.androidLogLevel ?: Log.ASSERT) <= Log.ASSERT) {
            try {
                formatString(message, parameters)?.let { Log.println(Log.ASSERT, LOGTAG, it) }
            } catch (e: Exception) {
                formatString(formatErrorMessage, message, parameters.contentToString())?.let {
                    Log.e(
                        LOGTAG,
                        it
                    )
                }
            }
        }
    }

    override fun lockLogLevel() {
        logLevelLocked = true
    }

    companion object {
        private const val formatErrorMessage = "Error formating log message: %s, with params: %s"
    }

    fun formatString(format: String?, vararg args: Any?): String? {
        if (format.isNullOrBlank()) {
            return "null"
        }

        return try {
            java.lang.String.format(Locale.US, format, args)
        } catch (e: Exception) {
            // Handle any formatting exceptions here (e.g., MissingFormatArgumentException)
            Log.e("formatString", "Formatting error: ${e.message}")
            null
        }
    }
}